File: python/arrays/nextPermutation.py
def nextPermutation(nums: list[int]):
    # find the pivot element
    pivot = -1

    for i in range(len(nums) - 1, 0, -1):
        if nums[i - 1] < nums[i]:
            pivot = i - 1
            break

    # if no pivot found just reverse the array and return
    if pivot == -1:
        nums[:] = nums[::-1]
        return

    # find the next smallest Largest value than pivot
    for i in range(len(nums) - 1, pivot, -1):
        if nums[i] > nums[pivot]:
            # when found swap them
            print(nums[i])
            nums[i], nums[pivot] = nums[pivot], nums[i]
            break

    # reverse the array after pivot
    nums[pivot + 1 :] = nums[-1:pivot:-1]


nums = [1, 3, 2]
nextPermutation(nums)
print(nums)

File: cpp/sorting/nextPermutation.cpp
#include "bits/stdc++.h"
#include <algorithm>
using namespace std;

void nextPermutation(vector<int> &nums) {

  // first find pivot element
  // pivot -> the element is < its right element starting from right
  // e.g [1,2,5,4,3] -> pivot -> 2 -> 2 < 5

  int n = nums.size();
  int i = n - 1;
  int pivotInd = 0;

  while (i > 0 && nums[i - 1] >= nums[i])
    i--;
  pivotInd = i - 1;

  if (pivotInd == -1) {
    // array is already in last permutation just reverse
    sort(nums.begin(), nums.end());
    return;
  }

  i = n - 1;
  int minLarInd = n - 1;

  while (i > pivotInd) {
    if (nums[i] > nums[pivotInd]) {
      minLarInd = i;
      break;
    }
    i--;
  }

  // swap pivot with next minimum largest
  swap(nums[pivotInd], nums[minLarInd]);
  sort(nums.begin() + pivotInd + 1, nums.end());
}

int main() {

  vector<int> nums = {2, 3, 1};

  nextPermutation(nums);

  for (int ele : nums) {
    cout << ele << " ";
  }

  return 0;
}
